### ISSUE TITLE
Change icon of terminal via argument / automatically to match running app (e.g. Vim)

### ISSUE THREAD

**Issue by oxysoft** [kovidgoyal/kitty#5503]
**Is your feature request related to a problem? Please describe.**
Hi, I run Vim in Kitty and even have an OS shortcut to directly open Kitty with vim inside. (where quitting Vim quits the whole terminal) When I am running commands in parallel in another terminal and start minimizing windows, I can easily get lost.

**Describe the solution you'd like**
It would be a very useful if the icon in the taskbar could be Vim's instead, since I'm really just using that terminal instance as a text editor rather than a shell. For starters, just manually specifying the icon with a launch argument would already be very useful for me, but if it's possible I'd love if the terminal could automatically update its icon to match what's running inside, since I also open vim temporarily some times. (might require the shell integration since you'd need to talk with the shell to know that, I think?) And of course, we'd keep the Kitty icon if the running app has no icon.

**Additional context**
Surprisingly I can't find much on this subject, even with other terminals. Some terminals can be specified through argument, but I think the automatic switch could be a great innovative feature for Kitty.

Cheers!

**Comment by kovidgoyal** [comment]
Not something I see much benefit in for the effort involved. One would
have to write cross platform code to get the icon of the foreground
program, then write cross platform code to change the icon of yourself,
which as far as I know is actually impossible on some platforms such as
Wayland (the only way to do it on Wayland is change app_id, which is
ridiculous).  

As for setting the icon at startup, you can do so already on Wayland by
using --class to change the app_id of the kitty instance to be the app
id of whatever program has the icon you want. On macOS you can simply
create a wrapper/launcher bundle that has whatever icon you like and
runs the kitty binary underneath. And X11 will have some program like
xprop that will allow you to write a script to change the icon of any
running program.

Finally, kitty is targeted at terminal power users, not people that don't
use the terminal other than as a wrapper for a single program, so I am
not very sympathetic to this use case.

**Comment by page-down** [comment]
kitty supports Linux and macOS, based on your description, I'm guessing you're using it under Linux.
On Linux, as far as I know, there is no standardized specification for setting the running app icon across different desktop environments.

> ... and start minimizing windows, I can easily get lost.
> ... I also open vim temporarily ...

I would suggest addressing the root cause of the problem. Use a more efficient way instead of looking at the icon and using the mouse to click on it.

For example, configure remote control for the desired kitty instances and filter the window ids from `kitty @ ls` (using `--to` to point to the remote control sockets of each running instance) via `fzf`, and switch the focus via `kitty @ focus-window`.

The following is an example of focusing a kitty window by filtering title. (Does not include filtering windows in multiple kitty instances)

```shell
kitty @ ls | jq -r '.[] | .tabs[] | .windows[] | select(.is_focused == false) | (.id|tostring) + " " + .title' | fzf --with-nth 2.. --bind 'enter:execute-silent(kitty @ focus-window -m id:{1})+accept'
```

As for the alternatives to displaying cli program icons:

For displaying icons on the tab bar, you can use custom tab bar with icon fonts. See the following discussion for some inspiration.
https://github.com/kovidgoyal/kitty/discussions/4447

If you want to display an icon on the window background, you can use `launch --logo`.
https://sw.kovidgoyal.net/kitty/launch/#cmdoption-launch-logo

![](https://user-images.githubusercontent.com/72330683/139504549-30245dfa-c46f-49b5-83ce-536062611174.png)

Then use any task switcher that has a thumbnail display to switch OS windows.

If you really need a graphical display of icons to switch OS windows, remote control protocol is available for desktop environments (e.g. KDE widgets), system trays (very easy to change icons), third-party TUI (kitty graphics protocol to display icons) programs, and GUI (including web-based) programs to call.

In short, what you are requesting requires support from all cross-platform desktop environments, which is currently not available.
